### PR TITLE
Feature/serial

### PR DESCRIPTION
**Pull request by dellaert**
A prototype implementation for serial manipulators.

If assumed massless (@danbarla ) the Jacobian calculated by the poe method can be used to compute torques $tau=J^T \mathcal{F}$ with F the wrench at the end-effector (foot). We can "easily" add joint angle dynamics, actually, in a later PR, I think.

The last test shows decomposition of the Panda, effortlessly.

### CODE REVIEW

File: python/tests/prototype/serial.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            return poe if fTe is None else poe.compose(fTe)
+        else:
+            # Compute FK + Jacobian with monoid compose.
+            assert J.shape == (6, len(q)), f"Needs 6x{len(q)} J."
+            Empty = np.zeros((6, 0))
+            pair = self.sMb, Empty
+            for j in range(n):
+                pair = compose(pair, (exp[j], np.expand_dims(A[:, j], 1)))
+            if fTe is not None:
+                pair = compose(pair, (fTe, Empty))  # Adjoints Jacobian to E!
```
Comment (dellaert) [inline python/tests/prototype/serial.py:136]: @Joobz this fixes the bug i alluded to.

File: python/tests/prototype/serial.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+"""
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ *
+ * @file  serial.py
+ * @brief Ptototype Serial class for serial manipulators.
```
Comment (Joobz) [inline python/tests/prototype/serial.py:8]: Typo: Ptototype to Prototype

File: python/tests/prototype/serial.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    """Monoid operation for pose,Jacobian pairs."""
+    aTb, Jb = A
+    bTc, Jc = B
+    assert Jb.shape[0] == 6 and Jc.shape[0] == 6, f"{Jb.shape} and {Jc.shape}"
+
+    # Compose poses
+    aTc = aTb.compose(bTc)
+
+    # Check if one of specs is an offset:
+    if Jb.shape[1] == 0:
```
Comment (Joobz) [inline python/tests/prototype/serial.py:29]: Are these ifs really needed? 
If Jb or Jc are empty, they will just not add any new columns when stacking. 
I see that for the Jb if, you evade computing the adjointmap, but the Jc one doesn't solve much.

File: python/tests/prototype/serial.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        # Convert all joints into pose/Jacobian pairs.
+        pairs = [cls(joint.pMc(), joint.cScrewAxis()) for joint in joints]
+
+        if base_name is not None:
+            # Create offset to first link parent.
+            assert joint_range is None or joint_range[0] == 0, \
+                "Cannot have base name if first joint is not 0"
+            base_link = robot.link(base_name)
+            sM0 = base_link.bMcom()
```
Comment (Joobz) [inline python/tests/prototype/serial.py:94]: I guess that the parent frame of the first joint is always the center of mass one? It just goes to show how I've never worked with GT dynamics 🙂

File: python/tests/prototype/serial.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            pairs = [offset] + pairs
+
+        # Now, let compose do the work!
+        return cls.compose(*pairs)
+
+    def poe(self, q: np.ndarray,
+            fTe: Optional[Pose3] = None,
+            J: Optional[np.ndarray] = None):
+        """ Perform forward kinematics given q, return Pose of end-effector.
+            When q is smaller than #joints, actuates last joints in chain.
```
Comment (Joobz) [inline python/tests/prototype/serial.py:105]: Maybe I am understanding it wrong, but I think this is contradictory with line 117 where it is checked that len(q) is equal to the number of joint axes considered.

File: python/tests/test_serial.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        ])
+        expected_sM7 = Pose3(sR7, Point3(0.0882972, 0.00213401, 0.933844))
+        actual_sM7 = gtd.Pose(fk, 7)
+        self.gtsamAssertEquals(actual_sM7, expected_sM7, tol=1e-3)
+
+        # Check that Serial can be constructed from robot
+        sM7, J7 = self.serial.spec()
+        self.gtsamAssertEquals(sM7, expected_sM7, tol=1e-3)
+
+        # Check Panda twsist in end-effector frame
```
Comment (Joobz) [inline python/tests/test_serial.py:132]: Typo: twsist to twist

File: python/tests/prototype/serial.py (hunk 1 above)
Comment (Joobz) [reply python/tests/prototype/serial.py:136]: Yay, I think I would have missed this one, totally flew over my head at first.

**Review by Joobz — COMMENTED** [review]
Awesome code. Just left some comments for doubts and 2 typos I noticed.

Also, first review!

File: python/tests/test_serial.py (hunk 6 above)
Comment (dellaert) [reply python/tests/test_serial.py:132]: Done

File: python/tests/prototype/serial.py (hunk 2 above)
Comment (dellaert) [reply python/tests/prototype/serial.py:8]: Done

File: python/tests/prototype/serial.py (hunk 3 above)
Comment (dellaert) [reply python/tests/prototype/serial.py:29]: Thanks, much simpler without checks and still works!

File: python/tests/prototype/serial.py (hunk 4 above)
Comment (dellaert) [reply python/tests/prototype/serial.py:94]: Yes

File: python/tests/prototype/serial.py (hunk 5 above)
Comment (dellaert) [reply python/tests/prototype/serial.py:105]: Good catch! Remove obsolete comment.

File: python/tests/prototype/serial.py (hunk 1 above)
Comment (dellaert) [reply python/tests/prototype/serial.py:136]: :-)

**Comment by dellaert** [comment]
@danbarla I added some math for you to read

**Comment by dellaert** [comment]
@Joobz also approve if you're satisfied I can easily address comments.

File: python/tests/prototype/chain.py [hunk 7]
```diff
@@ -16,29 +16,23 @@
...
-    """Monoid operation for pose,Jacobian pairs."""
-    aTb, Jb = A
-    bTc, Jc = B
-    assert Jb.shape[0] == 6 and Jc.shape[0] == 6, f"{Jb.shape} and {Jc.shape}"
+def compose(aSbj: Tuple[Pose3, np.ndarray], bSck: Tuple[Pose3, np.ndarray]):
+    """Monoid operation for chains, i.e., pose,Jacobian pairs."""
+    aTb, bAj = aSbj
+    bTc, cAk = bSck
+    assert bAj.shape[0] == 6 and cAk.shape[0] == 6,\
+        f"Jaocobains should have 6 rows, sapes are {bAj.shape} and {cAk.shape}"
```
Comment (Joobz) [inline python/tests/prototype/chain.py:24]: Typos: Jaocobains -> Jacobians and sapes -> shapes

**Review by Joobz — COMMENTED** [review]
I just found a typo. Everything else seems good.

File: python/tests/prototype/chain.py (hunk 7 above)
Comment (dellaert) [reply python/tests/prototype/chain.py:24]: Thanks! I'll take this as "approve" and merge. Note there is a button you should use to "approve" a PR :-)

Comment (Joobz) [reply python/tests/prototype/chain.py:24]: Oops, sorry about that, now I know :-)

**Pull request merged**